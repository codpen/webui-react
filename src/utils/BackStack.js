import Stack from './Stack'
// A very simple stack implementation to handle back links in remote explorer
class BackStack {
    constructor() {
        this.backStack = new Stack();
        this.forwardStack = new Stack();
    }

    getLength() {
        return this.backStack.getLength();
    }

    push(item) {

        this.backStack.push(item);
        this.forwardStack.empty();

    }

    pop() {
        const temp = this.backStack.pop();
        return temp;
    }

    peek() {
        return this.backStack.peek();
    }

    empty() {
        this.backStack.empty();
        this.forwardStack.empty();
    }

    moveBack() {
        const temp = this.backStack.moveBack();
        if(temp)
            this.forwardStack.push(temp);
        return temp;
    }

    moveForward(){
        const temp = this.forwardStack.pop();
        if(temp){
            //Pop was successful
            this.backStack.push(temp);
        }
        return temp;

    }



}

export default BackStack;