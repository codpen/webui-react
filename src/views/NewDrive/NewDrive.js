import React from 'react';
import {Button, Card, CardBody, CardFooter, CardHeader, Col, Collapse, Form, FormGroup, Input, Label} from "reactstrap";

import {config} from "./config.js";
import axios from 'axios';
import cloneDeep from 'lodash/cloneDeep';

function DriveParameters({driveType, loadAdvanced, changeHandler, currentValues}) {
    if (driveType !== undefined && driveType !== "") {
        const inputsMap = config[driveType].Options;
        console.log("current values" + currentValues);

        let outputMap = inputsMap.map((attr, idx) => {
            if ((loadAdvanced && attr.Advanced) || (!loadAdvanced && !attr.Advanced)) {
                return (
                    <FormGroup key={idx} row>
                        <Label for={attr.Name} sm={5}>{attr.Help}</Label>
                        <Col sm={7}>
                            <Input type={attr.IsPassword ? "password" : "text"} value={currentValues[attr.Name]}
                                   name={attr.Name}
                                   id={attr.Name} onChange={changeHandler}/>
                        </Col>
                    </FormGroup>
                );
            } else {
                return null;
            }
        });
        return outputMap;
    }
    return (
        <div>Select a drive type to continue</div>
    );
}

function DriveTypes(props) {
    let configMap = config.map((drive, idx) => (
        <option key={idx} value={idx}>{drive.Description}</option>
    ));

    return configMap;

}

function CustomInput({key, id, label, changeHandler, type, value, name, placeholder}) {
    return (
        <FormGroup key={key} row>
            <Label for={id} sm={5}>{label}</Label>
            <Col sm={7}>
                <Input type={type} value={value} name={name} placeholder={placeholder}
                       id={id} onChange={changeHandler}/>
            </Col>
        </FormGroup>);
}


class NewDrive extends React.Component {

    constructor(props, context) {
        super(props, context);
        this.state = {

            colRconfig: true,
            colSetup: false,
            colAdvanced: false,
            driveName: "",


            advancedOptions: false,
            formValues: {},


            driveType: ""
        };
        this.handleSubmit = this.handleSubmit.bind(this);
    }

    toggle = (e) => {
        let name = e.target.name;
        console.log("Name:" + name);

        this.setState({[name]: !this.state[name]})
    };

    handleInputChange = (e) => {
        let new_prod_diff = cloneDeep(this.state.formValues);
        new_prod_diff[e.target.name] = e.target.value;
        this.setState({formValues: new_prod_diff});

    };

    changeDriveType = (e) => {
        e.preventDefault();

        let val = e.target.value;

        let availableOptions = {};
        let drivePrefix = "";
        // console.log(config[val].Options);
        if (val !== undefined && val !== "") {

            config[val].Options.forEach(item => {
                availableOptions[item.Name] = item.Default;
            });
            drivePrefix = config[val].Prefix;
        }

        this.setState({driveType: val, drivePrefix: drivePrefix, formValues: availableOptions},
            (prevState) => {
                console.log(config[this.state.driveType]);
            });


    };

    openSetupDrive = (e) => {
        e.preventDefault();
        this.setState({'colSetup': true})
        this.setupDriveDiv.scrollIntoView({behavior: "smooth"});
    };

    editAdvancedOptions = (e) => {
        this.setState({advancedOptions: !this.state.advancedOptions});
        // console.log(e.target.value);
    };

    handleSubmit(e) {
        e.preventDefault();
        console.log("Submitted form");
        const headers = {
            headers: {'Content-Type': 'application/json'}
        };
        let data = {parameters: this.state.formValues, name: this.state.driveName, type: this.state.drivePrefix};

        axios.post('http://localhost:5572/config/create', data, headers).then(function (response) {
            console.log(response.data); // ex.: { user: 'Your User'}
            console.log(response.status); // ex.: 200
        });
    }

    changeName = e => {
        this.setState({driveName: e.target.value})
    };

    openAdvancedSettings = e => {
        if (this.state.advancedOptions) {
            this.setState({colAdvanced: true});
        } else {
            this.configEndDiv.scrollIntoView({behavior: "smooth"});
        }
    };


    render() {
        const {colRconfig, colSetup, colAdvanced, driveType, advancedOptions} = this.state;
        const {driveName} = this.state;
        return (
            <div>
                <p>This 3 step process will guide you through creating a new config. If you want </p>
                <Form onSubmit={this.handleSubmit}>
                    <Card>
                        <CardHeader>
                            <h5>
                                <Button color="link" name="colRconfig" onClick={this.toggle}
                                        style={{marginBottom: '1rem'}}><strong>Step 1:</strong> Remote Config</Button>
                            </h5>

                        </CardHeader>
                        <Collapse isOpen={colRconfig}>

                            <CardBody>
                                <CustomInput label="Name of this drive (For your reference)"
                                             changeHandler={this.changeName} value={driveName}
                                             placeholder={"Enter a name"} name="name" id="driveName"/>

                                <FormGroup row>
                                    <Label for="driveType" sm={5}>Select</Label>
                                    <Col sm={7}>
                                        <Input type="select" name="type" id="driveType" value={driveType}
                                               onChange={this.changeDriveType}>
                                            <option value="">Select one</option>
                                            <DriveTypes/>
                                        </Input>
                                    </Col>
                                </FormGroup>
                                <FormGroup row>
                                    <Col sm={3}>
                                        <Label for="inputDriveName">Docs are available at </Label>{' '}
                                        <a href="https://rclone.org/commands/rclone_config/">Rclone Config</a>
                                    </Col>


                                </FormGroup>
                            </CardBody>

                            <CardFooter>
                                <div className="clearfix">
                                    <Button color="success" className="float-right" onClick={this.openSetupDrive}><i
                                        className="fa fa-check fa-lg "/></Button>
                                </div>
                            </CardFooter>

                        </Collapse>
                    </Card>
                    <Card>
                        {/*div for Scrolling to here*/}
                        <div ref={(el) => this.setupDriveDiv = el}/>
                        <CardHeader>
                            <h5>
                                <Button color="link" name="colSetup" onClick={this.toggle}
                                        style={{marginBottom: '1rem'}}><strong>Step 2:</strong> Setup Drive</Button>
                            </h5>

                        </CardHeader>
                        <Collapse isOpen={colSetup}>

                            <CardBody>
                                <DriveParameters driveType={driveType} loadAdvanced={false}
                                                 changeHandler={this.handleInputChange}
                                                 currentValues={this.state.formValues}/>
                            </CardBody>
                            <CardFooter>
                                <div className="clearfix">
                                    <div className="float-right">
                                        <Input type="checkbox" value={advancedOptions}
                                               onChange={this.editAdvancedOptions}/><span className="mr-3">Edit Advanced Options</span>
                                        <Button color="success" onClick={this.openAdvancedSettings}><i
                                            className="fa fa-check fa-lg "/></Button>

                                    </div>
                                </div>
                            </CardFooter>

                        </Collapse>
                    </Card>
                    <Card>
                        <CardHeader>
                            <h5>
                                <Button color="link" name="colAdvanced" onClick={this.toggle}
                                        style={{marginBottom: '1rem'}}><strong>Step 3:</strong> Advanced
                                    (optional)</Button>

                            </h5>

                        </CardHeader>
                        <Collapse isOpen={colAdvanced}>

                            <CardBody>
                                <DriveParameters driveType={driveType} loadAdvanced={true}
                                                 changeHandler={this.handleInputChange}
                                                 currentValues={this.state.formValues}/>
                            </CardBody>
                            <CardFooter>
                                <div className="clearfix">
                                    <Button color="success" className="float-right"><i className="fa fa-check fa-lg "/></Button>
                                </div>
                            </CardFooter>

                        </Collapse>
                    </Card>
                    <div className="clearfix" ref={(el) => {
                        this.configEndDiv = el
                    }}>
                        <div className="float-right mb-3">
                            <Button color="success" type="submit">Create Config</Button>

                        </div>
                    </div>
                </Form>
            </div>);
    }
}

export default NewDrive;
