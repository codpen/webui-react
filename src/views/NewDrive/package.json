{
  "name": "MyDashboard",
  "version": "0.0.0",
  "private": true,
  "main": "NewDrive.js"
}
