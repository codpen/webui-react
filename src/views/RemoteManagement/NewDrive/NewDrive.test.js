import React from "react";
import {shallow} from "enzyme";
import {findByTestAttr, testStore} from "../../../../Utils";
import NewDrive from "./NewDrive";
import toJson from "enzyme-to-json";
import {TEST_REDUX_PROPS} from "../../../utils/testData";

const setUp = (intialState = {}, props = {}) => {
    const store = testStore(intialState);
    const component = shallow(<NewDrive {...props} store={store}/>);
    return component.childAt(0).dive();
};


describe('Remote Explorer', function () {


    describe('renders', function () {
        let wrapper;
        beforeEach(() => {
            const initialState = {
                ...TEST_REDUX_PROPS
            };


            const props = {
                match: {
                    params: {
                        drivePrefix: ""
                    }
                }
            };
            wrapper = setUp(initialState, props)
        });

        it('should render without crashing', function () {
            const component = findByTestAttr(wrapper, "newDriveComponent");
            expect(component).toHaveLength(1);
        });
        it('should match snapshot', function () {
            expect(toJson(wrapper)).toMatchSnapshot()
        });
    });

});