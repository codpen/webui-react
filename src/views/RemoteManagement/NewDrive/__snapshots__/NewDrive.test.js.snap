// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Remote Explorer renders should match snapshot 1`] = `
<div
  data-test="newDriveComponent"
>
  <p>
    This 3 step process will guide you through creating a new config. For auto config, leave the parameters as is.
  </p>
  <Form
    onSubmit={[Function]}
    tag="form"
  >
    <Card
      tag="div"
    >
      <CardHeader
        tag="div"
      >
        <h5>
          <Button
            color="link"
            name="colRconfig"
            onClick={[Function]}
            style={
              Object {
                "marginBottom": "1rem",
              }
            }
            tag="button"
          >
            <strong>
              Step 1:
            </strong>
             Remote Config
          </Button>
        </h5>
      </CardHeader>
      <Collapse
        appear={false}
        enter={true}
        exit={true}
        in={false}
        isOpen={true}
        mountOnEnter={false}
        onEnter={[Function]}
        onEntered={[Function]}
        onEntering={[Function]}
        onExit={[Function]}
        onExited={[Function]}
        onExiting={[Function]}
        tag="div"
        timeout={350}
        unmountOnExit={false}
      >
        <CardBody
          tag="div"
        >
          <CustomInput
            changeHandler={[Function]}
            id="driveName"
            isValid={false}
            label="Name of this drive (For your reference)"
            name="name"
            placeholder="Enter a name"
            value=""
          />
          <FormGroup
            row={true}
            tag="div"
          >
            <Label
              for="driveType"
              sm={5}
              tag="label"
              widths={
                Array [
                  "xs",
                  "sm",
                  "md",
                  "lg",
                  "xl",
                ]
              }
            >
              Select
            </Label>
            <Col
              sm={7}
              tag="div"
              widths={
                Array [
                  "xs",
                  "sm",
                  "md",
                  "lg",
                  "xl",
                ]
              }
            >
              <ProviderAutoSuggest
                onChange={[Function]}
                suggestions={
                  Array [
                    Object {
                      "Description": "Backblaze B2",
                      "Name": "b2",
                      "Options": Array [
                        Object {
                          "Advanced": false,
                          "Default": "",
                          "DefaultStr": "",
                          "Help": "Account ID or Application Key ID",
                          "Hide": 0,
                          "IsPassword": false,
                          "Name": "account",
                          "NoPrefix": false,
                          "Provider": "",
                          "Required": true,
                          "ShortOpt": "",
                          "Type": "string",
                          "Value": null,
                          "ValueStr": "",
                        },
                        Object {
                          "Advanced": false,
                          "Default": "",
                          "DefaultStr": "",
                          "Help": "Application Key",
                          "Hide": 0,
                          "IsPassword": false,
                          "Name": "key",
                          "NoPrefix": false,
                          "Provider": "",
                          "Required": true,
                          "ShortOpt": "",
                          "Type": "string",
                          "Value": null,
                          "ValueStr": "",
                        },
                        Object {
                          "Advanced": true,
                          "Default": "",
                          "DefaultStr": "",
                          "Help": "Endpoint for the service.
Leave blank normally.",
                          "Hide": 0,
                          "IsPassword": false,
                          "Name": "endpoint",
                          "NoPrefix": false,
                          "Provider": "",
                          "Required": false,
                          "ShortOpt": "",
                          "Type": "string",
                          "Value": null,
                          "ValueStr": "",
                        },
                        Object {
                          "Advanced": true,
                          "Default": "",
                          "DefaultStr": "",
                          "Help": "A flag string for X-Bz-Test-Mode header for debugging.

This is for debugging purposes only. Setting it to one of the strings
below will cause b2 to return specific errors:

  * \\"fail_some_uploads\\"
  * \\"expire_some_account_authorization_tokens\\"
  * \\"force_cap_exceeded\\"

These will be set in the \\"X-Bz-Test-Mode\\" header which is documented
in the [b2 integrations checklist](https://www.backblaze.com/b2/docs/integration_checklist.html).",
                          "Hide": 2,
                          "IsPassword": false,
                          "Name": "test_mode",
                          "NoPrefix": false,
                          "Provider": "",
                          "Required": false,
                          "ShortOpt": "",
                          "Type": "string",
                          "Value": null,
                          "ValueStr": "",
                        },
                        Object {
                          "Advanced": true,
                          "Default": false,
                          "DefaultStr": "false",
                          "Help": "Include old versions in directory listings.
Note that when using this no file write operations are permitted,
so you can't upload files or delete them.",
                          "Hide": 0,
                          "IsPassword": false,
                          "Name": "versions",
                          "NoPrefix": false,
                          "Provider": "",
                          "Required": false,
                          "ShortOpt": "",
                          "Type": "bool",
                          "Value": null,
                          "ValueStr": "false",
                        },
                        Object {
                          "Advanced": false,
                          "Default": false,
                          "DefaultStr": "false",
                          "Help": "Permanently delete files on remote removal, otherwise hide files.",
                          "Hide": 0,
                          "IsPassword": false,
                          "Name": "hard_delete",
                          "NoPrefix": false,
                          "Provider": "",
                          "Required": false,
                          "ShortOpt": "",
                          "Type": "bool",
                          "Value": null,
                          "ValueStr": "false",
                        },
                        Object {
                          "Advanced": true,
                          "Default": 209715200,
                          "DefaultStr": "200M",
                          "Help": "Cutoff for switching to chunked upload.

Files above this size will be uploaded in chunks of \\"--b2-chunk-size\\".

This value should be set no larger than 4.657GiB (== 5GB).",
                          "Hide": 0,
                          "IsPassword": false,
                          "Name": "upload_cutoff",
                          "NoPrefix": false,
                          "Provider": "",
                          "Required": false,
                          "ShortOpt": "",
                          "Type": "SizeSuffix",
                          "Value": null,
                          "ValueStr": "200M",
                        },
                        Object {
                          "Advanced": true,
                          "Default": 100663296,
                          "DefaultStr": "96M",
                          "Help": "Upload chunk size. Must fit in memory.

When uploading large files, chunk the file into this size.  Note that
these chunks are buffered in memory and there might a maximum of
\\"--transfers\\" chunks in progress at once.  5,000,000 Bytes is the
minimum size.",
                          "Hide": 0,
                          "IsPassword": false,
                          "Name": "chunk_size",
                          "NoPrefix": false,
                          "Provider": "",
                          "Required": false,
                          "ShortOpt": "",
                          "Type": "SizeSuffix",
                          "Value": null,
                          "ValueStr": "96M",
                        },
                        Object {
                          "Advanced": true,
                          "Default": false,
                          "DefaultStr": "false",
                          "Help": "Disable checksums for large (> upload cutoff) files",
                          "Hide": 0,
                          "IsPassword": false,
                          "Name": "disable_checksum",
                          "NoPrefix": false,
                          "Provider": "",
                          "Required": false,
                          "ShortOpt": "",
                          "Type": "bool",
                          "Value": null,
                          "ValueStr": "false",
                        },
                        Object {
                          "Advanced": true,
                          "Default": "",
                          "DefaultStr": "",
                          "Help": "Custom endpoint for downloads.

This is usually set to a Cloudflare CDN URL as Backblaze offers
free egress for data downloaded through the Cloudflare network.
This is probably only useful for a public bucket.
Leave blank if you want to use the endpoint provided by Backblaze.",
                          "Hide": 0,
                          "IsPassword": false,
                          "Name": "download_url",
                          "NoPrefix": false,
                          "Provider": "",
                          "Required": false,
                          "ShortOpt": "",
                          "Type": "string",
                          "Value": null,
                          "ValueStr": "",
                        },
                        Object {
                          "Advanced": true,
                          "Default": 604800000000000,
                          "DefaultStr": "1w",
                          "Help": "Time before the authorization token will expire in s or suffix ms|s|m|h|d.

The duration before the download authorization token will expire.
The minimum value is 1 second. The maximum value is one week.",
                          "Hide": 0,
                          "IsPassword": false,
                          "Name": "download_auth_duration",
                          "NoPrefix": false,
                          "Provider": "",
                          "Required": false,
                          "ShortOpt": "",
                          "Type": "Duration",
                          "Value": null,
                          "ValueStr": "1w",
                        },
                      ],
                      "Prefix": "b2",
                    },
                  ]
                }
                value=""
              />
            </Col>
          </FormGroup>
          <FormGroup
            row={true}
            tag="div"
          >
            <Col
              sm={3}
              tag="div"
              widths={
                Array [
                  "xs",
                  "sm",
                  "md",
                  "lg",
                  "xl",
                ]
              }
            >
              <Label
                for="inputDriveName"
                tag="label"
                widths={
                  Array [
                    "xs",
                    "sm",
                    "md",
                    "lg",
                    "xl",
                  ]
                }
              >
                Docs are available at 
              </Label>
               
              <a
                href="https://rclone.org/commands/rclone_config/"
              >
                Rclone Config
              </a>
            </Col>
          </FormGroup>
        </CardBody>
        <CardFooter
          tag="div"
        >
          <div
            className="clearfix"
          >
            <Button
              className="float-right"
              color="success"
              onClick={[Function]}
              tag="button"
            >
              <i
                className="fa fa-check fa-lg "
              />
            </Button>
          </div>
        </CardFooter>
      </Collapse>
    </Card>
    <Card
      tag="div"
    >
      <div />
      <CardHeader
        tag="div"
      >
        <h5>
          <Button
            color="link"
            name="colSetup"
            onClick={[Function]}
            style={
              Object {
                "marginBottom": "1rem",
              }
            }
            tag="button"
          >
            <strong>
              Step 2:
            </strong>
             Setup Drive
          </Button>
        </h5>
      </CardHeader>
      <Collapse
        appear={false}
        enter={true}
        exit={true}
        in={false}
        isOpen={false}
        mountOnEnter={false}
        onEnter={[Function]}
        onEntered={[Function]}
        onEntering={[Function]}
        onExit={[Function]}
        onExited={[Function]}
        onExiting={[Function]}
        tag="div"
        timeout={350}
        unmountOnExit={false}
      >
        <CardBody
          tag="div"
        >
          <DriveParameters
            changeHandler={[Function]}
            config={
              Array [
                Object {
                  "Description": "Backblaze B2",
                  "Name": "b2",
                  "Options": Array [
                    Object {
                      "Advanced": false,
                      "Default": "",
                      "DefaultStr": "",
                      "Help": "Account ID or Application Key ID",
                      "Hide": 0,
                      "IsPassword": false,
                      "Name": "account",
                      "NoPrefix": false,
                      "Provider": "",
                      "Required": true,
                      "ShortOpt": "",
                      "Type": "string",
                      "Value": null,
                      "ValueStr": "",
                    },
                    Object {
                      "Advanced": false,
                      "Default": "",
                      "DefaultStr": "",
                      "Help": "Application Key",
                      "Hide": 0,
                      "IsPassword": false,
                      "Name": "key",
                      "NoPrefix": false,
                      "Provider": "",
                      "Required": true,
                      "ShortOpt": "",
                      "Type": "string",
                      "Value": null,
                      "ValueStr": "",
                    },
                    Object {
                      "Advanced": true,
                      "Default": "",
                      "DefaultStr": "",
                      "Help": "Endpoint for the service.
Leave blank normally.",
                      "Hide": 0,
                      "IsPassword": false,
                      "Name": "endpoint",
                      "NoPrefix": false,
                      "Provider": "",
                      "Required": false,
                      "ShortOpt": "",
                      "Type": "string",
                      "Value": null,
                      "ValueStr": "",
                    },
                    Object {
                      "Advanced": true,
                      "Default": "",
                      "DefaultStr": "",
                      "Help": "A flag string for X-Bz-Test-Mode header for debugging.

This is for debugging purposes only. Setting it to one of the strings
below will cause b2 to return specific errors:

  * \\"fail_some_uploads\\"
  * \\"expire_some_account_authorization_tokens\\"
  * \\"force_cap_exceeded\\"

These will be set in the \\"X-Bz-Test-Mode\\" header which is documented
in the [b2 integrations checklist](https://www.backblaze.com/b2/docs/integration_checklist.html).",
                      "Hide": 2,
                      "IsPassword": false,
                      "Name": "test_mode",
                      "NoPrefix": false,
                      "Provider": "",
                      "Required": false,
                      "ShortOpt": "",
                      "Type": "string",
                      "Value": null,
                      "ValueStr": "",
                    },
                    Object {
                      "Advanced": true,
                      "Default": false,
                      "DefaultStr": "false",
                      "Help": "Include old versions in directory listings.
Note that when using this no file write operations are permitted,
so you can't upload files or delete them.",
                      "Hide": 0,
                      "IsPassword": false,
                      "Name": "versions",
                      "NoPrefix": false,
                      "Provider": "",
                      "Required": false,
                      "ShortOpt": "",
                      "Type": "bool",
                      "Value": null,
                      "ValueStr": "false",
                    },
                    Object {
                      "Advanced": false,
                      "Default": false,
                      "DefaultStr": "false",
                      "Help": "Permanently delete files on remote removal, otherwise hide files.",
                      "Hide": 0,
                      "IsPassword": false,
                      "Name": "hard_delete",
                      "NoPrefix": false,
                      "Provider": "",
                      "Required": false,
                      "ShortOpt": "",
                      "Type": "bool",
                      "Value": null,
                      "ValueStr": "false",
                    },
                    Object {
                      "Advanced": true,
                      "Default": 209715200,
                      "DefaultStr": "200M",
                      "Help": "Cutoff for switching to chunked upload.

Files above this size will be uploaded in chunks of \\"--b2-chunk-size\\".

This value should be set no larger than 4.657GiB (== 5GB).",
                      "Hide": 0,
                      "IsPassword": false,
                      "Name": "upload_cutoff",
                      "NoPrefix": false,
                      "Provider": "",
                      "Required": false,
                      "ShortOpt": "",
                      "Type": "SizeSuffix",
                      "Value": null,
                      "ValueStr": "200M",
                    },
                    Object {
                      "Advanced": true,
                      "Default": 100663296,
                      "DefaultStr": "96M",
                      "Help": "Upload chunk size. Must fit in memory.

When uploading large files, chunk the file into this size.  Note that
these chunks are buffered in memory and there might a maximum of
\\"--transfers\\" chunks in progress at once.  5,000,000 Bytes is the
minimum size.",
                      "Hide": 0,
                      "IsPassword": false,
                      "Name": "chunk_size",
                      "NoPrefix": false,
                      "Provider": "",
                      "Required": false,
                      "ShortOpt": "",
                      "Type": "SizeSuffix",
                      "Value": null,
                      "ValueStr": "96M",
                    },
                    Object {
                      "Advanced": true,
                      "Default": false,
                      "DefaultStr": "false",
                      "Help": "Disable checksums for large (> upload cutoff) files",
                      "Hide": 0,
                      "IsPassword": false,
                      "Name": "disable_checksum",
                      "NoPrefix": false,
                      "Provider": "",
                      "Required": false,
                      "ShortOpt": "",
                      "Type": "bool",
                      "Value": null,
                      "ValueStr": "false",
                    },
                    Object {
                      "Advanced": true,
                      "Default": "",
                      "DefaultStr": "",
                      "Help": "Custom endpoint for downloads.

This is usually set to a Cloudflare CDN URL as Backblaze offers
free egress for data downloaded through the Cloudflare network.
This is probably only useful for a public bucket.
Leave blank if you want to use the endpoint provided by Backblaze.",
                      "Hide": 0,
                      "IsPassword": false,
                      "Name": "download_url",
                      "NoPrefix": false,
                      "Provider": "",
                      "Required": false,
                      "ShortOpt": "",
                      "Type": "string",
                      "Value": null,
                      "ValueStr": "",
                    },
                    Object {
                      "Advanced": true,
                      "Default": 604800000000000,
                      "DefaultStr": "1w",
                      "Help": "Time before the authorization token will expire in s or suffix ms|s|m|h|d.

The duration before the download authorization token will expire.
The minimum value is 1 second. The maximum value is one week.",
                      "Hide": 0,
                      "IsPassword": false,
                      "Name": "download_auth_duration",
                      "NoPrefix": false,
                      "Provider": "",
                      "Required": false,
                      "ShortOpt": "",
                      "Type": "Duration",
                      "Value": null,
                      "ValueStr": "1w",
                    },
                  ],
                  "Prefix": "b2",
                },
              ]
            }
            currentValues={Object {}}
            drivePrefix=""
            errorsMap={
              Object {
                "driveName": "",
              }
            }
            isValidMap={Object {}}
            loadAdvanced={false}
          />
        </CardBody>
        <CardFooter
          tag="div"
        >
          <div
            className="clearfix"
          >
            <div
              className="float-right"
            >
              <Input
                onChange={[Function]}
                type="checkbox"
                value={false}
              />
              <span
                className="mr-3"
              >
                Edit Advanced Options
              </span>
              <Button
                color="success"
                onClick={[Function]}
                tag="button"
              >
                <i
                  className="fa fa-check fa-lg "
                />
              </Button>
            </div>
          </div>
        </CardFooter>
      </Collapse>
    </Card>
    <Card
      tag="div"
    >
      <CardHeader
        tag="div"
      >
        <h5>
          <Button
            color="link"
            name="colAdvanced"
            onClick={[Function]}
            style={
              Object {
                "marginBottom": "1rem",
              }
            }
            tag="button"
          >
            <strong>
              Step 3:
            </strong>
             Advanced (optional)
          </Button>
        </h5>
      </CardHeader>
      <Collapse
        appear={false}
        enter={true}
        exit={true}
        in={false}
        isOpen={false}
        mountOnEnter={false}
        onEnter={[Function]}
        onEntered={[Function]}
        onEntering={[Function]}
        onExit={[Function]}
        onExited={[Function]}
        onExiting={[Function]}
        tag="div"
        timeout={350}
        unmountOnExit={false}
      >
        <CardBody
          tag="div"
        >
          <DriveParameters
            changeHandler={[Function]}
            config={
              Array [
                Object {
                  "Description": "Backblaze B2",
                  "Name": "b2",
                  "Options": Array [
                    Object {
                      "Advanced": false,
                      "Default": "",
                      "DefaultStr": "",
                      "Help": "Account ID or Application Key ID",
                      "Hide": 0,
                      "IsPassword": false,
                      "Name": "account",
                      "NoPrefix": false,
                      "Provider": "",
                      "Required": true,
                      "ShortOpt": "",
                      "Type": "string",
                      "Value": null,
                      "ValueStr": "",
                    },
                    Object {
                      "Advanced": false,
                      "Default": "",
                      "DefaultStr": "",
                      "Help": "Application Key",
                      "Hide": 0,
                      "IsPassword": false,
                      "Name": "key",
                      "NoPrefix": false,
                      "Provider": "",
                      "Required": true,
                      "ShortOpt": "",
                      "Type": "string",
                      "Value": null,
                      "ValueStr": "",
                    },
                    Object {
                      "Advanced": true,
                      "Default": "",
                      "DefaultStr": "",
                      "Help": "Endpoint for the service.
Leave blank normally.",
                      "Hide": 0,
                      "IsPassword": false,
                      "Name": "endpoint",
                      "NoPrefix": false,
                      "Provider": "",
                      "Required": false,
                      "ShortOpt": "",
                      "Type": "string",
                      "Value": null,
                      "ValueStr": "",
                    },
                    Object {
                      "Advanced": true,
                      "Default": "",
                      "DefaultStr": "",
                      "Help": "A flag string for X-Bz-Test-Mode header for debugging.

This is for debugging purposes only. Setting it to one of the strings
below will cause b2 to return specific errors:

  * \\"fail_some_uploads\\"
  * \\"expire_some_account_authorization_tokens\\"
  * \\"force_cap_exceeded\\"

These will be set in the \\"X-Bz-Test-Mode\\" header which is documented
in the [b2 integrations checklist](https://www.backblaze.com/b2/docs/integration_checklist.html).",
                      "Hide": 2,
                      "IsPassword": false,
                      "Name": "test_mode",
                      "NoPrefix": false,
                      "Provider": "",
                      "Required": false,
                      "ShortOpt": "",
                      "Type": "string",
                      "Value": null,
                      "ValueStr": "",
                    },
                    Object {
                      "Advanced": true,
                      "Default": false,
                      "DefaultStr": "false",
                      "Help": "Include old versions in directory listings.
Note that when using this no file write operations are permitted,
so you can't upload files or delete them.",
                      "Hide": 0,
                      "IsPassword": false,
                      "Name": "versions",
                      "NoPrefix": false,
                      "Provider": "",
                      "Required": false,
                      "ShortOpt": "",
                      "Type": "bool",
                      "Value": null,
                      "ValueStr": "false",
                    },
                    Object {
                      "Advanced": false,
                      "Default": false,
                      "DefaultStr": "false",
                      "Help": "Permanently delete files on remote removal, otherwise hide files.",
                      "Hide": 0,
                      "IsPassword": false,
                      "Name": "hard_delete",
                      "NoPrefix": false,
                      "Provider": "",
                      "Required": false,
                      "ShortOpt": "",
                      "Type": "bool",
                      "Value": null,
                      "ValueStr": "false",
                    },
                    Object {
                      "Advanced": true,
                      "Default": 209715200,
                      "DefaultStr": "200M",
                      "Help": "Cutoff for switching to chunked upload.

Files above this size will be uploaded in chunks of \\"--b2-chunk-size\\".

This value should be set no larger than 4.657GiB (== 5GB).",
                      "Hide": 0,
                      "IsPassword": false,
                      "Name": "upload_cutoff",
                      "NoPrefix": false,
                      "Provider": "",
                      "Required": false,
                      "ShortOpt": "",
                      "Type": "SizeSuffix",
                      "Value": null,
                      "ValueStr": "200M",
                    },
                    Object {
                      "Advanced": true,
                      "Default": 100663296,
                      "DefaultStr": "96M",
                      "Help": "Upload chunk size. Must fit in memory.

When uploading large files, chunk the file into this size.  Note that
these chunks are buffered in memory and there might a maximum of
\\"--transfers\\" chunks in progress at once.  5,000,000 Bytes is the
minimum size.",
                      "Hide": 0,
                      "IsPassword": false,
                      "Name": "chunk_size",
                      "NoPrefix": false,
                      "Provider": "",
                      "Required": false,
                      "ShortOpt": "",
                      "Type": "SizeSuffix",
                      "Value": null,
                      "ValueStr": "96M",
                    },
                    Object {
                      "Advanced": true,
                      "Default": false,
                      "DefaultStr": "false",
                      "Help": "Disable checksums for large (> upload cutoff) files",
                      "Hide": 0,
                      "IsPassword": false,
                      "Name": "disable_checksum",
                      "NoPrefix": false,
                      "Provider": "",
                      "Required": false,
                      "ShortOpt": "",
                      "Type": "bool",
                      "Value": null,
                      "ValueStr": "false",
                    },
                    Object {
                      "Advanced": true,
                      "Default": "",
                      "DefaultStr": "",
                      "Help": "Custom endpoint for downloads.

This is usually set to a Cloudflare CDN URL as Backblaze offers
free egress for data downloaded through the Cloudflare network.
This is probably only useful for a public bucket.
Leave blank if you want to use the endpoint provided by Backblaze.",
                      "Hide": 0,
                      "IsPassword": false,
                      "Name": "download_url",
                      "NoPrefix": false,
                      "Provider": "",
                      "Required": false,
                      "ShortOpt": "",
                      "Type": "string",
                      "Value": null,
                      "ValueStr": "",
                    },
                    Object {
                      "Advanced": true,
                      "Default": 604800000000000,
                      "DefaultStr": "1w",
                      "Help": "Time before the authorization token will expire in s or suffix ms|s|m|h|d.

The duration before the download authorization token will expire.
The minimum value is 1 second. The maximum value is one week.",
                      "Hide": 0,
                      "IsPassword": false,
                      "Name": "download_auth_duration",
                      "NoPrefix": false,
                      "Provider": "",
                      "Required": false,
                      "ShortOpt": "",
                      "Type": "Duration",
                      "Value": null,
                      "ValueStr": "1w",
                    },
                  ],
                  "Prefix": "b2",
                },
              ]
            }
            currentValues={Object {}}
            drivePrefix=""
            errorsMap={
              Object {
                "driveName": "",
              }
            }
            isValidMap={Object {}}
            loadAdvanced={true}
          />
        </CardBody>
      </Collapse>
    </Card>
    <div
      className="clearfix"
    >
      <div
        className="float-right mb-3"
      >
        <Button
          color="info"
          onClick={[Function]}
          tag="button"
          type="reset"
        >
          Clear
        </Button>
        <Button
          color="success"
          tag="button"
          type="submit"
        >
          Create Config
        </Button>
      </div>
    </div>
  </Form>
  <NewDriveAuthModal
    closeModal={[Function]}
    isVisible={false}
  />
</div>
`;
